File: backend/app/schemas/tasks.py
from datetime import datetime
from enum import Enum
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


# ── Enums ─────────────────────────────────────────────────────────────────────


class TaskPriority(str, Enum):
    low      = "low"
    medium   = "medium"
    high     = "high"
    critical = "critical"


class TaskStatus(str, Enum):
    pending     = "pending"
    in_progress = "in_progress"
    completed   = "completed"
    cancelled   = "cancelled"


class TaskFocusType(str, Enum):
    pomodoro  = "pomodoro"
    stopwatch = "stopwatch"
    none      = "none"


# ── Embedded ──────────────────────────────────────────────────────────────────


class TaskTagResponse(BaseModel):
    id:    UUID
    name:  str
    color: str


# ── Request schemas ───────────────────────────────────────────────────────────


class TaskCreate(BaseModel):
    title:       str            = Field(..., min_length=1, max_length=255, strip_whitespace=True)
    description: str | None    = Field(None, max_length=2000)
    priority:    TaskPriority   = TaskPriority.medium
    focus_type:  TaskFocusType  = TaskFocusType.none
    category_id: UUID | None   = None
    due_date:    datetime | None = None
    # Tags to attach at creation time; validated against the user's own tags.
    tag_ids:     list[UUID]    = Field(default_factory=list)


class TaskUpdate(BaseModel):
    title:       str | None    = Field(None, min_length=1, max_length=255, strip_whitespace=True)
    description: str | None    = None
    priority:    TaskPriority | None  = None
    # Completed/cancelled transitions are handled by dedicated PATCH endpoints
    # which also trigger side-effects (impact calculation, etc.).
    status:      Literal["pending", "in_progress"] | None = None
    focus_type:  TaskFocusType | None = None
    category_id: UUID | None   = None   # send null to clear
    due_date:    datetime | None = None  # send null to clear


class AddTagsRequest(BaseModel):
    tag_ids: list[UUID] = Field(..., min_length=1)


# ── Response schemas ──────────────────────────────────────────────────────────


class TaskResponse(BaseModel):
    id:                        UUID
    user_id:                   UUID
    category_id:               UUID | None
    title:                     str
    description:               str | None
    priority:                  TaskPriority
    status:                    TaskStatus
    focus_type:                TaskFocusType
    due_date:                  datetime | None
    completed_at:              datetime | None
    estimated_adventure_impact: int | None
    # Always populated — never return a task without its tags.
    tags:                      list[TaskTagResponse]
    created_at:                datetime
    updated_at:                datetime


class TaskHistoryResponse(BaseModel):
    tasks:  list[TaskResponse]
    total:  int
    limit:  int
    offset: int

File: web/src/features/tasks/schemas/tasks.ts
import { z } from "zod"

export const TaskTagResponseSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  color: z.string(),
})

export const TaskResponseSchema = z.object({
  id: z.string().uuid(),
  user_id: z.string().uuid(),
  category_id: z.string().uuid().nullable().optional(),
  title: z.string(),
  description: z.string().nullable().optional(),
  priority: z.enum(["low", "medium", "high", "critical"]),
  status: z.enum(["pending", "in_progress", "completed", "cancelled"]),
  focus_type: z.enum(["pomodoro", "stopwatch", "none"]),
  due_date: z.string().nullable().optional(),
  completed_at: z.string().nullable().optional(),
  estimated_adventure_impact: z.number().nullable().optional(),
  tags: z.array(TaskTagResponseSchema),
  created_at: z.string(),
  updated_at: z.string(),
})

export const TaskCreateSchema = z.object({
  title: z.string().min(1).max(255),
  description: z.string().nullable().optional(),
  priority: z.enum(["low", "medium", "high", "critical"]).optional(),
  focus_type: z.enum(["pomodoro", "stopwatch", "none"]).optional(),
  category_id: z.string().uuid().nullable().optional(),
  due_date: z.string().nullable().optional(),
  // Accept either an array of UUIDs or a comma-separated string (frontend convenience)
  tag_ids: z.union([z.array(z.string().uuid()), z.string()]).nullable().optional(),
})

export const TaskUpdateSchema = TaskCreateSchema.extend({
  status: z.enum(["pending", "in_progress"]).nullable().optional(),
})

export type TaskResponseType = z.infer<typeof TaskResponseSchema>
export type TaskCreateType = z.infer<typeof TaskCreateSchema>
export type TaskUpdateType = z.infer<typeof TaskUpdateSchema>
